connectors/stash/src/main/java/com/vmware/connectors/stash/StashController.java

```diff
@@ -0,0 +1,444 @@
+            consumes = MediaType.APPLICATION_FORM_URLENCODED_VALUE
+    )
+    public Single<ResponseEntity<String>> decline(final StashPullRequest pullRequest,
+                                                @RequestHeader(AUTH_HEADER) final String authHeader,
+                                                @RequestHeader(BASE_URL_HEADER) final String baseUrl) {
+
+        logger.debug("Decline action for stash pull request: {}, baseURL: {}",
+                pullRequest,
+                baseUrl);
+
+        return performStashAction(baseUrl, authHeader, pullRequest, StashAction.DECLINE);
+    }
+
+    @PostMapping(
+            path = "/api/v1/{projectKey}/{repositorySlug}/{pullRequestId}/comments",
+            consumes = MediaType.APPLICATION_FORM_URLENCODED_VALUE
+    )
+    public Single<ResponseEntity<String>> comments(final StashPullRequest pullRequest,
+                                                  @RequestParam(COMMENT_PARAM_KEY) final String comment,
+                                                  @RequestHeader(AUTH_HEADER) final String authHeader,
+                                                  @RequestHeader(BASE_URL_HEADER) final String baseUrl) {
+
+        logger.debug("Comment action for stash pull request: {}, baseURL: {}",
+                pullRequest,
+                baseUrl);
+
+        final HttpHeaders headers = buildHeaders(authHeader);
+        headers.add(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE);
+
+        final String url = String.format(StashConstants.STASH_PR_COMMENT_URL_FORMAT,
+                baseUrl,
+                pullRequest.getProjectKey(),
+                pullRequest.getRepositorySlug(),
+                pullRequest.getPullRequestId());
+
+        final StashComment stashComment = new StashComment(comment);
+
+        final ListenableFuture<ResponseEntity<String>> response = this.rest.exchange(
+                url,
+                HttpMethod.POST,
+                new HttpEntity<>(stashComment, headers),
+                String.class);
+        return Async.toSingle(response);
+    }
+
+    private List<String> getComments(final String baseUrl,
+                                     final HttpHeaders headers,
+                                     final StashPullRequest pullRequest) {
+        final String url = String.format(StashConstants.STASH_ACTIVITIES_URL_FORMAT,
+                baseUrl,
+                pullRequest.getProjectKey(),
+                pullRequest.getRepositorySlug(),
+                pullRequest.getPullRequestId());
+
+        final ListenableFuture<ResponseEntity<JsonDocument>> response = this.rest.exchange(
+                url,
+                HttpMethod.GET,
+                new HttpEntity<>(headers),
+                JsonDocument.class);
+
+        final JsonDocument[] jsonDocuments = new JsonDocument[1];
+        Async.toSingle(response)
+                .subscribe(entity -> {
+                    jsonDocuments[0] = entity.getBody();
```

Comment:
I'm confused by this.  Isn't this asynchronous and will fall through to the Assert before jsonDocuments[0] is set?

If it happens to be retrieving the comments before falling through, we probably shouldn't be handling this how it's currently written.

We should instead make this in the same asynchronous style as elsewhere.  I'll give a more concrete suggestion after I finish reviewing the rest of the code.

Reply:
What Ryan said. You almost never want to call `subscribe`.

Reply:
Confirmed that this pattern isn't working:

```json
{
  "results": [
    {
      "connector_id": "c5f0d317-9770-42c2-bf53-4102145a5ea9",
      "connector_status": {
        "status": 500,
        "connector_message": "{\"timestamp\":\"2018-01-26T21:36:11.284+0000\",\"status\":500,\"error\":\"Internal Server Error\",\"exception\":\"java.lang.IllegalArgumentException\",\"message\":\"JsonDocument should not be null.\",\"path\":\"/cards/requests\"}"
      }
    }
  ]
}
```

To test, create a connector for this with:

```
curl -v \
     -X POST \
     -H 'Content-Type: application/json' \
     -H "Authorization: Bearer $ADMIN_JWT_TOKEN" \
     -d '
{
    "name": "Stash",
    "external_connector_discovery_url": "http://172.17.100.100:8069",
    "backend_base_url": "https://stash.air-watch.com",
    "authentication": {
        "authentication_type": "BASIC"
    }
}
' \
     "http://$HOST:$PORT/tenantconfig/tenant-connectors"
```

On my box, api-gateway is running on localhost:8080, so $HOST is localhost, $PORT is 8080.  For my external discovery url, I have a vm on 172.17.100.100 (you can create one too with: vagrant up), where I cloned the connectors repo, build with the make-rpm and install that rpm (my application.properties sets the port to 8069).  You can simplify this on your box if you just run the connector with spring boot:run or something.

Whatever ip+port you run the connector on, put in the curl's external_connector_discovery_url.

With that configured, I ran the following cards request to get the above error:

```
curl -v \
     -X POST \
     -H "Content-Type: application/json" \
     -H "Accept: application/json" \
     -H "Authorization: Bearer $USER_JWT_TOKEN" \
     -H "X-Stash-Authorization: Basic $jira_creds" \
     -d "
{
    \"card_requests\": [
        {
            \"tokens\": {
                \"pr_email_subject\": [
                    \"UFO/app-platform-server - Pull request #252: APF-767 Loosen up handling of connector error responses\"
                ]
            },
            \"connector_id\": \"$stash_connector_id\"
        }
    ]
}
" \
     "http://$HOST:$PORT/hero/cards/requests"
```

Where $stash_connector_id is the connector id you can get from discovery and $jira_creds is "user:password" base64 encoded.  Here is a script to prompt and output that (to prevent you from accidentally putting your password in bash history or something):

```
#!/bin/bash

read -p 'User: ' user
read -s -p 'Password: ' password
echo ''
echo -n "$user:$password" | base64
```

Remember that this blob of text is sensitive (your password), so if you store it in a file, don't share it and make sure the read permissions on it are strict.  If this ends up in an environment variable (or logs), be sure to clean up afterwards.

Reply:
$ADMIN_JWT_TOKEN is tenant admin token
$USER_JWT_TOKEN is the user's token

I'll send links to our wiki page w/ dev tokens in chat.

Reply:
You probably have to do something like this:

```java
    private Single<List<String>> getComments(final String baseUrl,
                                     final HttpHeaders headers,
                                     final StashPullRequest pullRequest) {
        final String url = String.format(StashConstants.STASH_ACTIVITIES_URL_FORMAT,
                baseUrl,
                pullRequest.getProjectKey(),
                pullRequest.getRepositorySlug(),
                pullRequest.getPullRequestId());

        final ListenableFuture<ResponseEntity<JsonDocument>> response = this.rest.exchange(
                url,
                HttpMethod.GET,
                new HttpEntity<>(headers),
                JsonDocument.class);

        return Async.toSingle(response)
                .map(ResponseEntity::getBody)
                .map(jsonDocument -> jsonDocument.<List<String>>read("$.values[*].comment.text"))
                .map(comments -> comments.stream().limit(COMMENTS_SIZE).collect(Collectors.toList()));
    }

    private Observable<Card> getCardsForStashPR(final HttpHeaders headers,
                                                final StashPullRequest pullRequest,
                                                final String baseUrl,
                                                final String routingPrefix) {
        logger.debug("Requesting pull request info from stash base url: {} and pull request info: {}", baseUrl, pullRequest);

        final Single<ResponseEntity<JsonDocument>> stashResponse = Async.toSingle(getPullRequestInfo(headers, pullRequest, baseUrl));
        final Single<List<String>> comments = getComments(baseUrl, headers, pullRequest);

        return Observable.zip(stashResponse.toObservable(), comments.toObservable(), Pair::of)
                .onErrorResumeNext(ObservableUtil::skip404)
                .map(pair -> convertResponseIntoCard(pair.getLeft(), pullRequest, routingPrefix, pair.getRight()));
    }
```

Reply:
I have verified that Observable.zip is what you want to use.

Reply:
Changed the code to above format.. Thanks for the suggestion... Since I have not tested the code initially, I did not realize the mistake... Tested it now and it works now...